src/mlpack/core/util/using.hpp

```diff
@@ -71,6 +72,49 @@ namespace mlpack {
 
 #endif
 
+  namespace internal_compact {
+
+    namespace fill {
+
+      #ifdef MLPACK_HAS_COOT
+        struct fill_none  : public arma::fill::fill_class<
+                              arma::fill::fill_none>,
```

Comment:
I strongly advise against directly using `arma::fill::fill_class<>`.  This is an implementation level detail that can be changed in future versions of Armadillo.

Only the following symbols are publicly documented within Armadillo docs: `arma::fill::none`, `arma::fill::zeros`, `arma::fill::ones`, `arma::fill::eye`, `arma::fill::randu`, `arma::fill::randn`, `arma::fill::value`.

Extract from the on [API stability section](https://arma.sourceforge.net/docs.html#changelog) of Armadillo documentation:  
> ...     Any functionality within Armadillo which is not explicitly described in the public API documentation is considered as internal implementation detail, and may be changed or removed without notice.

In other words, `arma::fill::fill_class<>` is off-limits.

Reply:
@conradsnicta this one is tricky. I agree and would prefer not to use internal Armadillo types, but here we want one name we can use that can be *both* `arma::fill::zeros` and `coot::fill::zeros`.  This helps us avoid the situation where mlpack developers can't use `fill::`, which would feel like an awkward restriction.  Ideally writing internal mlpack code feels as much like writing Armadillo as possible, except to keep it general we avoid the `arma::`.  Without some namespace-agnostic solution for `fill::`, we can't have any uninitialized matrices, which is sometimes desired. (i.e. avoid filling the matrix with zeros, since the next call fills it.)

Under those desiderata, the only solution I could come up with is inheriting from the base class underlying both `arma::fill::zeros` and `coot::fill::zeros`. Do you have any alternative suggestions?  I'm out today so I don't think I'll be able to play with it until Monday. Maybe there is some clever trick that C++ allows that we haven't thought of yet.

Reply:
Maybe we can cleverly use `decltype` to avoid the direct name `arma::fill::fill_class`? We would still be depending on the fact that `arma::fill::zeros` is a type and not an enum.

Reply:
:facepalm: oh, I was way overthinking it, yeah, a simple template metaprogramming approach like that would work great.  It's a couple more characters, but that's not an issue.

> PS. I'm still not convinced that abstracting away the matrix type is the right way forward, though that discussion should probably be continued in the contributors-chat repo. All of this has the smell of overengineering.

I see where you're coming from, but I also don't see another reasonable way to allow mlpack to work on matrix types that aren't supported directly by Armadillo.